Math/NumberTheory/EisensteinIntegers.hs

```diff
@@ -0,0 +1,421 @@
+-- |
+-- Module:      Math.NumberTheory.EisensteinIntegers
+-- Copyright:   (c) 2018 Alexandre Rodrigues Baldé
+-- Licence:     MIT
+-- Maintainer:  Alexandre Rodrigues Baldé <alexandrer_b@outlook.com>
+-- Stability:   Provisional
+-- Portability: Non-portable (GHC extensions)
+--
+-- This module exports functions for manipulating Eisenstein integers, including
+-- computing their prime factorisations.
+--
+
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE BinaryLiterals #-}
+{-# LANGUAGE DeriveGeneric  #-}
+{-# LANGUAGE RankNTypes     #-}
+
+module Math.NumberTheory.EisensteinIntegers
+  ( EisensteinInteger(..)
+  , ω
+  , conjugate
+  , norm
+  , associates
+  , ids
+
+  -- * Division and remainder functions
+  , divE
+  , divModE
+  , modE
+  , quotRemE
+  , quotE
+  , remE
+
+  , divideByThree
+  , gcdE
+
+  -- * Primality functions
+  , factorise
+  , findPrime
+  , isPrime
+  , primes
+  ) where
+
+import Data.Bits                                       ((.&.), setBit, shift,
+                                                        zeroBits)
+import Data.List                                       (mapAccumL, partition)
+import Data.Maybe                                      (fromMaybe)
+import Data.Ord                                        (comparing)
+import GHC.Generics                                    (Generic)
+
+import qualified Math.NumberTheory.Moduli               as Moduli
+import Math.NumberTheory.Moduli.Sqrt                    (FieldCharacteristic(..))
+import qualified Math.NumberTheory.Primes.Factorisation as Factorisation
+import Math.NumberTheory.Primes.Types                   (PrimeNat(..))
+import qualified Math.NumberTheory.Primes.Sieve         as Sieve
+import qualified Math.NumberTheory.Primes.Testing       as Testing
+import Math.NumberTheory.Utils                          (mergeBy)
+import Math.NumberTheory.Utils.FromIntegral             (integerToNatural)
+
+infix 6 :+
+
+-- | An Eisenstein integer is a + bω, where a and b are both integers.
+data EisensteinInteger = (:+) { real :: !Integer, imag :: !Integer }
+    deriving (Eq, Ord, Generic)
+
+-- | The imaginary unit for Eisenstein integers, where
+--
+-- > ω == (-1/2) + ((sqrt 3)/2)ι == exp(2*pi*ι/3)
+-- and ι is the usual imaginary unit with ι² == -1.
+ω :: EisensteinInteger
+ω = 0 :+ 1
+
+instance Show EisensteinInteger where
+    show (a :+ b)
+        | b == 0     = show a
+        | a == 0     = s ++ b'
+        | otherwise  = show a ++ op ++ b'
+        where
+            b' = if abs b == 1 then "ω" else show (abs b) ++ "*ω"
+            op = if b > 0 then "+" else "-"
+            s  = if b > 0 then "" else "-"
+
+instance Num EisensteinInteger where
+    (+) (a :+ b) (c :+ d) = (a + c) :+ (b + d)
+    (*) (a :+ b) (c :+ d) = (a * c - b * d) :+ (b * c + a * d - b * d)
```

Comment:
It is worth to save one multiplication: `b * (c - d) + a * d` or `b * c + (a - b) * d`.